File: app/java/src/main/java/com/example/production/domain/model/subcontract/SupplyType.java
package com.example.production.domain.model.subcontract;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

/**
 * 支給区分
 */
@Getter
@RequiredArgsConstructor
public enum SupplyType {
    PAID("有償支給"),
    FREE("無償支給");

    private final String displayName;

    public static SupplyType fromDisplayName(String displayName) {
        for (SupplyType type : values()) {
            if (type.displayName.equals(displayName)) {
                return type;
            }
        }
        throw new IllegalArgumentException("Unknown supply type: " + displayName);
    }
}

File: app/csharp/src/ProductionManagement.Domain/Models/Subcontract/SupplyType.cs
namespace ProductionManagement.Domain.Models.Subcontract;

/// <summary>
/// 支給区分
/// </summary>
public enum SupplyType
{
    /// <summary>有償支給</summary>
    Paid,
    /// <summary>無償支給</summary>
    Free
}

/// <summary>
/// SupplyType 拡張メソッド
/// </summary>
public static class SupplyTypeExtensions
{
    private static readonly Dictionary<SupplyType, string> DisplayNames = new()
    {
        { SupplyType.Paid, "有償支給" },
        { SupplyType.Free, "無償支給" }
    };

    private static readonly Dictionary<string, SupplyType> FromDisplayNames =
        DisplayNames.ToDictionary(x => x.Value, x => x.Key);

    public static string GetDisplayName(this SupplyType type) => DisplayNames[type];

    public static SupplyType FromDisplayName(string displayName)
    {
        if (FromDisplayNames.TryGetValue(displayName, out var type))
        {
            return type;
        }
        throw new ArgumentException($"Unknown supply type: {displayName}");
    }
}
